### ISSUE TITLE
About training time / Slow training schedule

### ISSUE THREAD

**Issue by AlphaGoMK** [NVlabs/wetectron#25]
Hi, thanks for your great work. I am going to reproduce your results but encounter long training time issue. I've tried on 8 M40 GPU, and process would finish after more than 2 days 15 hours later. I'm not sure whether something go wrong?  
Could you provide your training log and detailed environments?  
I train this repo using command `python -m torch.distributed.launch --nproc_per_node=8 tools/train_net.py --config-file "configs/voc/V_16_voc07.yaml" OUTPUT_DIR ./experiment_voc`, and pytorch==1.5.1, cudatoolkit==10.1.243  
Hope you could help me figure out the issue, thanks in advance.

**Comment by jason718** [comment]
I don't have access to M40, not sure how fast it is. I was using 8 V100, and the training took 9~10h for voc07.

**Comment by jshi31** [comment]
I met the same issue. The estimated training time for COCO is 6 days with batch size 8 on 4 GPUs. I further test the forward time with batch size 1 for training, which costs 0.3s for each iteration on average. 
The timing bottleneck turns out to be:
1. The unique() operation cost on average 0.2s, which is really wired. 
https://github.com/NVlabs/wetectron/blob/44e6fa95aee07d6722a62af56f016a3ae99bd8a6/wetectron/modeling/roi_heads/weak_head/loss.py#L249

2. The second is the mist_layer (where NMS is applied) which cost 0.1s. https://github.com/NVlabs/wetectron/blob/44e6fa95aee07d6722a62af56f016a3ae99bd8a6/wetectron/modeling/roi_heads/weak_head/loss.py#L258

These two time bottlenecks are linear to batch size, as they are computed for each image.
The second bottleneck is reasonable as NMS is time-consuming.
So I wonder if the first bottleneck only happens for me or for others.

Thanks

**Comment by AlphaGoMK** [comment]
> I met the same issue. The estimated training time for COCO is 6 days with batch size 8 on 4 GPUs. I further test the forward time with batch size 1 for training, which costs 0.3s for each iteration on average.
> The timing bottleneck turns out to be:
> 
> 1. The unique() operation cost on average 0.2s, which is really wired.
>    https://github.com/NVlabs/wetectron/blob/44e6fa95aee07d6722a62af56f016a3ae99bd8a6/wetectron/modeling/roi_heads/weak_head/loss.py#L249
> 2. The second is the mist_layer (where NMS is applied) which cost 0.1s. https://github.com/NVlabs/wetectron/blob/44e6fa95aee07d6722a62af56f016a3ae99bd8a6/wetectron/modeling/roi_heads/weak_head/loss.py#L258
> 
> These two time bottlenecks are linear to batch size, as they are computed for each image.
> The second bottleneck is reasonable as NMS is time-consuming.
> So I wonder if the first bottleneck only happens for me or for others.
> 
> Thanks

I find the same phenomenon as yours, and unique is more time-consuming in my machine.  
To tackle this, I cache the unique_labels in target by modifing the `__getitem__` method of voc dataset, which slightly improve the training speed.  

Another bottleneck is Concrete Drop Block, as shown below  
https://github.com/NVlabs/wetectron/blob/master/wetectron/engine/trainer.py#L216-L223  
CDB method process the whole network twice and update the parameters. Disable CDB will boost training speed about 2x, but this may harm model performance as mentioned in paper.  

After applied the above two modification, the training time reduced from 2d15h to about 20h on 8xM40  

And have you ever reproduced the results in paper? I've trained MIST w/ CDB on 4x2 GPUs, and the gap between my results and paper seems to large (51.12 VS 54.9).  
I wonder if I get something wrong in training?  

Thanks

**Comment by jason718** [comment]
somehow `unique()` wasn't causing significant slow-down for me. Can you guys share the profile record? If that were the issue, maybe we can move it to pre-processing or data fetching stage.

**Comment by AlphaGoMK** [comment]
> somehow `unique()` wasn't causing significant slow-down for me. Can you guys share the profile record? If that were the issue, maybe we can move it to pre-processing or data fetching stage.

My profile record is
```
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
      aten::_unique2        70.55%     119.751us       100.00%     169.730us     169.730us     122.688us        72.59%     169.024us     169.024us             1  
         aten::clone         6.95%      11.796us        24.01%      40.748us      40.748us      18.624us        11.02%      40.192us      40.192us             1  
    aten::empty_like         2.46%       4.171us         4.51%       7.652us       7.652us       0.000us         0.00%       0.000us       0.000us             1  
         aten::empty         2.05%       3.481us         2.05%       3.481us       3.481us       0.000us         0.00%       0.000us       0.000us             1  
         aten::copy_        12.55%      21.300us        12.55%      21.300us      21.300us      21.568us        12.76%      21.568us      21.568us             1  
       aten::reshape         2.89%       4.901us         3.70%       6.286us       6.286us       6.144us         3.63%       6.144us       6.144us             1  
          aten::view         0.82%       1.385us         0.82%       1.385us       1.385us       0.000us         0.00%       0.000us       0.000us             1  
         aten::empty         0.93%       1.573us         0.93%       1.573us       1.573us       0.000us         0.00%       0.000us       0.000us             1  
         aten::empty         0.37%       0.625us         0.37%       0.625us       0.625us       0.000us         0.00%       0.000us       0.000us             1  
       aten::resize_         0.44%       0.747us         0.44%       0.747us       0.747us       0.000us         0.00%       0.000us       0.000us             1  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 169.730us
CUDA time total: 169.024us
 0.1203913688659668
```

And my test code is
```
t0 = time.time()
with torch.autograd.profiler.profile(use_cuda=True) as prof:
    labels_per_im = targets_per_im.get_field('labels').unique()
print(prof, time.time()-t0)
```
Modified around https://github.com/NVlabs/wetectron/blob/master/wetectron/modeling/roi_heads/weak_head/loss.py#L249

**Comment by jshi31** [comment]
@jason718 Hi, could you share the log and training time for COCO? Now I am trying to replicate the performance of COCO, but the training time is 3 days and the performance is behind the reported number. I am wondering if it is because 
the default setting for coco in https://github.com/NVlabs/wetectron/blob/master/configs/coco/V_16_coco17.yaml does not use the concrete dropblock.

**Comment by jason718** [comment]
> > somehow `unique()` wasn't causing significant slow-down for me. Can you guys share the profile record? If that were the issue, maybe we can move it to pre-processing or data fetching stage.
> 
> My profile record is
> 
> ```
> --------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
>                 Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
> --------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
>       aten::_unique2        70.55%     119.751us       100.00%     169.730us     169.730us     122.688us        72.59%     169.024us     169.024us             1  
>          aten::clone         6.95%      11.796us        24.01%      40.748us      40.748us      18.624us        11.02%      40.192us      40.192us             1  
>     aten::empty_like         2.46%       4.171us         4.51%       7.652us       7.652us       0.000us         0.00%       0.000us       0.000us             1  
>          aten::empty         2.05%       3.481us         2.05%       3.481us       3.481us       0.000us         0.00%       0.000us       0.000us             1  
>          aten::copy_        12.55%      21.300us        12.55%      21.300us      21.300us      21.568us        12.76%      21.568us      21.568us             1  
>        aten::reshape         2.89%       4.901us         3.70%       6.286us       6.286us       6.144us         3.63%       6.144us       6.144us             1  
>           aten::view         0.82%       1.385us         0.82%       1.385us       1.385us       0.000us         0.00%       0.000us       0.000us             1  
>          aten::empty         0.93%       1.573us         0.93%       1.573us       1.573us       0.000us         0.00%       0.000us       0.000us             1  
>          aten::empty         0.37%       0.625us         0.37%       0.625us       0.625us       0.000us         0.00%       0.000us       0.000us             1  
>        aten::resize_         0.44%       0.747us         0.44%       0.747us       0.747us       0.000us         0.00%       0.000us       0.000us             1  
> --------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
> Self CPU time total: 169.730us
> CUDA time total: 169.024us
>  0.1203913688659668
> ```
> 
> And my test code is
> 
> ```
> t0 = time.time()
> with torch.autograd.profiler.profile(use_cuda=True) as prof:
>     labels_per_im = targets_per_im.get_field('labels').unique()
> print(prof, time.time()-t0)
> ```
> 
> Modified around https://github.com/NVlabs/wetectron/blob/master/wetectron/modeling/roi_heads/weak_head/loss.py#L249

This looks weird, the profile shows the running time is ~170us, but the printed time is  0.12s. But I don't see anything wrong with the `unique()` function. 170us seems to be reasonable.

**Comment by jason718** [comment]
> @jason718 Hi, could you share the log and training time for COCO? Now I am trying to replicate the performance of COCO, but the training time is 3 days and the performance is behind the reported number. I am wondering if it is because
> the default setting for coco in https://github.com/NVlabs/wetectron/blob/master/configs/coco/V_16_coco17.yaml does not use the concrete dropblock.

Yes I do have the log. Can you open another issue and report your running setting? It's hard for me to say whether "3 days" is good or bad without knowing your env.
